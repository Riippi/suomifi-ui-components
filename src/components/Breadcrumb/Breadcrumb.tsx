import React, { Component, ReactNode } from 'react';
import { HtmlNav, HtmlNavProps, HtmlOl, HtmlLi } from '../../reset';
import classnames from 'classnames';

export const baseClassName = 'fi-breadcrumb';
const listClassName = `${baseClassName}_list`;
const itemClassName = `${baseClassName}_item`;

export interface BreadcrumbProps extends HtmlNavProps {
  /** Name for the breadcrumb like 'Breadcrumb' */
  'aria-label': string;
  /** Custom classname to extend or customize */
  className?: string;
  /**
   * Link element displayed content
   */
  children?: ReactNode;
}

const breadcrumbItems = (children: ReactNode) =>
  React.Children.map(children, (child) => (
    <HtmlLi className={itemClassName}>{child}</HtmlLi>
  ));

export class Breadcrumb extends Component<BreadcrumbProps> {
  render() {
    const { className, children, ...passProps } = this.props;
    return (
      <HtmlNav {...passProps} className={classnames(baseClassName, className)}>
        <HtmlOl className={listClassName}>{breadcrumbItems(children)}</HtmlOl>
      </HtmlNav>
    );
  }
}
