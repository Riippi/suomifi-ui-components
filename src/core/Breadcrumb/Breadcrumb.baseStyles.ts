import { css } from 'styled-components';
import { suomifiTheme } from '../theme';
import { BreadcrumbProps } from './Breadcrumb';
import { nav, list, listItem, fonts } from '../theme/reset';

export const baseStyles = ({ theme = suomifiTheme }: BreadcrumbProps) => css`
  ${nav(theme)}
  ${fonts(theme).body}
  background-color: ${theme.colors.whiteBase};

  & .fi-breadcrumb {
    &_list {
      ${list(theme)}
      ${fonts(theme).body}
      margin: 0;
      padding: 0;
    }
    &_item {
      ${listItem(theme)}
      ${fonts(theme).body}
      float: left;
    }
    &_item,
    &_link,
    &_icon {
      font-size: ${theme.typography.fontSize.body};
    }
    &_icon {
      transform: translateY(.2em);
    }
  }
`;
