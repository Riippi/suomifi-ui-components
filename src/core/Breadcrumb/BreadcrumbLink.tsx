import React, { Fragment } from 'react';
import { ThemeProp, suomifiTheme } from '../theme';
import { baseClassName } from '../../components/Breadcrumb/Breadcrumb';
import { Link, LinkProps } from '../Link/Link';
import { Icon } from '../Icon/Icon';
import { HtmlSpan } from '../../reset';
import classnames from 'classnames';
import { Omit } from '../../utils/typescript';

const linkClassName = `${baseClassName}_link`;
const iconClassName = `${baseClassName}_icon`;

export interface BreadcrumbLinkProps extends Omit<LinkProps, 'href'> {
  /** Indicating the link is the current page */
  current?: boolean;
  /** url for the link */
  href?: string;
  theme?: ThemeProp;
}

export const BreadcrumbLink = ({
  current,
  children,
  className,
  href = '',
  ...passProps
}: BreadcrumbLinkProps) =>
  !!current ? (
    <HtmlSpan className={className} aria-current="location">
      {children}
    </HtmlSpan>
  ) : (
    <Fragment>
      <Link
        {...passProps}
        className={classnames(linkClassName, className)}
        href={href}
      >
        {children}
      </Link>
      <Icon
        icon="linkBreadcrumb"
        className={iconClassName}
        color={
          !!passProps.theme
            ? passProps.theme.colors.blackBase
            : suomifiTheme.colors.blackBase // TODO
        }
      />
    </Fragment>
  );
