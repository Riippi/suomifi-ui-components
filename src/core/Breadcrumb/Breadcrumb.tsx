import React, { Component } from 'react';
import { default as styled } from 'styled-components';
import { withDefaultTheme } from '../theme/utils';
import { ThemeComponent } from '../theme';
import { baseStyles } from './Breadcrumb.baseStyles';
import {
  Breadcrumb as CompBreadcrumb,
  BreadcrumbProps as CompBreadcrumbProps,
} from '../../components/Breadcrumb/Breadcrumb';
import { LinkProps } from '../Link/Link';
import { BreadcrumbLink, BreadcrumbLinkProps } from './BreadcrumbLink';

type BreadcrumbVariant = 'default' | 'link';

export interface BreadcrumbProps extends CompBreadcrumbProps, ThemeComponent {
  /**
   * 'default' | 'expansion'
   * @default default
   */
  variant?: BreadcrumbVariant;
}

const StyledBreadcrumb = styled(({ theme, ...passProps }: BreadcrumbProps) => (
  <CompBreadcrumb {...passProps} />
))`
  ${props => baseStyles(props)};
`;

type VariantBreadcrumbProps =
  | BreadcrumbProps
  | LinkProps & { variant?: BreadcrumbVariant };

/**
 * <i class="semantics" />
 * Used for navigation path
 */
export class Breadcrumb extends Component<VariantBreadcrumbProps> {
  static link = (props: BreadcrumbLinkProps) => {
    return <BreadcrumbLink {...withDefaultTheme(props)} />;
  };

  render() {
    const { variant, 'aria-label': ariaLabel, ...passProps } = withDefaultTheme(
      this.props,
    );
    if (variant === 'link') {
      return <BreadcrumbLink {...passProps as BreadcrumbLinkProps} />;
    }
    return (
      <StyledBreadcrumb
        {...passProps}
        aria-label={!!ariaLabel ? ariaLabel : 'breadcrumb'}
      />
    );
  }
}
