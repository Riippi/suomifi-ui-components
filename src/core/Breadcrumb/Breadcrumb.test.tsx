import React from 'react';
import { render } from 'react-testing-library';
import { axeTest } from '../../utils/test/axe';

import { Breadcrumb } from './Breadcrumb';

const TestBreadcrumb = (
  <Breadcrumb data-testid="Breadcrumb" aria-label="breadcrumb">
    <Breadcrumb.link href="/">Frontpage</Breadcrumb.link>
    <Breadcrumb.link href="/sub">Sub page</Breadcrumb.link>
    <Breadcrumb.link current={true}>Sub sub page</Breadcrumb.link>
  </Breadcrumb>
);

test('calling render with the same component on the same container does not remount', () => {
  const BreadcrumbRendered = render(TestBreadcrumb);
  const { container } = BreadcrumbRendered;
  expect(container.firstChild).toMatchSnapshot();
});

test('should not have basic accessibility issues', axeTest(TestBreadcrumb));
