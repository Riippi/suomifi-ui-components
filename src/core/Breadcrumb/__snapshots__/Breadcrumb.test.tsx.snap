// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.emotion-9 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  color: hsl(0,0%,16%);
  display: block;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial,sans-serif;
  font-size: 18px;
  font-weight: 400;
  line-height: 1.5;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  background-color: hsl(0,0%,100%);
}

.emotion-9 .fi-breadcrumb_list {
  color: hsl(0,0%,16%);
  list-style: none;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial,sans-serif;
  font-size: 18px;
  font-weight: 400;
  line-height: 1.5;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  margin: 0;
  padding: 0;
}

.emotion-9 .fi-breadcrumb_item {
  color: hsl(0,0%,16%);
  list-style: none;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial,sans-serif;
  font-size: 18px;
  font-weight: 400;
  line-height: 1.5;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  float: left;
}

.emotion-9 .fi-breadcrumb_item,
.emotion-9 .fi-breadcrumb_link,
.emotion-9 .fi-breadcrumb_icon {
  font-size: 18px;
}

.emotion-9 .fi-breadcrumb_icon {
  -webkit-transform: translateY(.2em);
  -ms-transform: translateY(.2em);
  transform: translateY(.2em);
}

.emotion-8 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  list-style-type: decimal;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 0;
  margin-right: 0;
  padding-left: 40px;
}

.emotion-2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: list-item;
}

.emotion-0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: hsl(0,0%,16%);
  font-family: 'Source Sans Pro','Helvetica Neue',Arial,sans-serif;
  font-size: 18px;
  font-weight: 400;
  line-height: 1.5;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0:focus,
.emotion-0:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.emotion-0:focus {
  outline: 0;
  position: relative;
}

.emotion-0:focus:after {
  content: '';
  position: absolute;
  top: -4px;
  right: -4px;
  bottom: -4px;
  left: -4px;
  border-radius: 4px;
  background-color: transparent;
  border: 1px solid hsl(23,82%,53%);
  box-sizing: border-box;
  box-shadow: 0 0 10px 0 hsl(23,82%,53%);
  z-index: 9999;
}

.emotion-0:focus:not(:focus-visible):after {
  content: none;
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0:focus,
.emotion-0:focus-within {
  color: hsl(212,63%,45%);
}

.emotion-0:hover,
.emotion-0:active,
.emotion-0:focus,
.emotion-0:focus-within {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:visited {
  color: hsl(284,36%,45%);
}

.emotion-1 {
  display: inline-block;
  vertical-align: baseline;
  -webkit-transform: translateY(0.1em);
  -ms-transform: translateY(0.1em);
  transform: translateY(0.1em);
}

.emotion-6 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

<nav
  aria-label="breadcrumb"
  class="fi-breadcrumb emotion-9"
  data-testid="Breadcrumb"
>
  <ol
    class="fi-breadcrumb_list emotion-8"
  >
    <li
      class="fi-breadcrumb_item emotion-2"
    >
      <a
        class="fi-breadcrumb_link fi-link emotion-0"
        href="/"
      >
        Frontpage
      </a>
      <svg
        aria-hidden="true"
        class="fi-breadcrumb_icon emotion-1"
        focusable="false"
        height="1em"
        style="fill: hsl(0, 0%, 16%);"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M6 18v-2.12l5.902-2.328 3.748-1.492v-.12l-3.748-1.492L6 8.119V6l12 5.015v1.97z"
        />
      </svg>
    </li>
    <li
      class="fi-breadcrumb_item emotion-2"
    >
      <a
        class="fi-breadcrumb_link fi-link emotion-0"
        href="/sub"
      >
        Sub page
      </a>
      <svg
        aria-hidden="true"
        class="fi-breadcrumb_icon emotion-1"
        focusable="false"
        height="1em"
        style="fill: hsl(0, 0%, 16%);"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M6 18v-2.12l5.902-2.328 3.748-1.492v-.12l-3.748-1.492L6 8.119V6l12 5.015v1.97z"
        />
      </svg>
    </li>
    <li
      class="fi-breadcrumb_item emotion-2"
    >
      <span
        aria-current="page"
        class="emotion-6"
      >
        Sub sub page
      </span>
    </li>
  </ol>
</nav>
`;
